File: transport/client/src/pacing/PacketPacer.ts
import { timerManager } from 'opencord-utils';
import type { LossEstimator } from '../fec';

const PACING_INTERVAL_MS = 5;
const LOSS_THRESHOLD = 0.10;
const MIN_INTERVALS = 3;
const MAX_INTERVALS = 6;
const MIN_BUDGET_BYTES = 10000;

export class PacketPacer {
  private queue: Uint8Array[] = [];
  private queueBytes: number = 0;
  private intervalId: number | null = null;
  private sendFn: (data: Uint8Array) => void;
  private lossEstimator: LossEstimator;

  constructor(
    sendFn: (data: Uint8Array) => void,
    lossEstimator: LossEstimator
  ) {
    this.sendFn = sendFn;
    this.lossEstimator = lossEstimator;
  }

  start(): void {
    if (this.intervalId !== null) return;
    this.intervalId = timerManager.setInterval(() => this.drain(), PACING_INTERVAL_MS);
  }

  stop(): void {
    if (this.intervalId !== null) {
      timerManager.clearInterval(this.intervalId);
      this.intervalId = null;
    }
    this.queue = [];
    this.queueBytes = 0;
  }

  enqueue(data: Uint8Array): void {
    this.queue.push(data);
    this.queueBytes += data.length;
  }

  private drain(): void {
    if (this.queue.length === 0) return;

    const budget = this.getBudget();
    let sent = 0;

    while (this.queue.length > 0 && sent + this.queue[0].length <= budget) {
      const packet = this.queue.shift()!;
      sent += packet.length;
      this.queueBytes -= packet.length;
      this.sendFn(packet);
    }
  }

  private getBudget(): number {
    const { lossRate } = this.lossEstimator.getStats();

    const t = Math.min(lossRate / LOSS_THRESHOLD, 1);
    const targetIntervals = MIN_INTERVALS + t * (MAX_INTERVALS - MIN_INTERVALS);
    const adaptiveBudget = this.queueBytes / targetIntervals;

    return Math.max(adaptiveBudget, MIN_BUDGET_BYTES);
  }

  get pending(): number {
    return this.queue.length;
  }

  reset(): void {
    this.queue = [];
    this.queueBytes = 0;
  }
}

File: transport/server/src/pacing/packet_pacer.rs
use bytes::Bytes;
use std::collections::VecDeque;

use crate::fec::LossEstimator;

const LOSS_THRESHOLD: f64 = 0.10;
const MIN_INTERVALS: f64 = 3.0;
const MAX_INTERVALS: f64 = 6.0;
const MIN_BUDGET_BYTES: usize = 10000;

pub struct PacketPacer {
    queue: VecDeque<Bytes>,
    queue_bytes: usize,
}

impl PacketPacer {
    pub fn new() -> Self {
        Self {
            queue: VecDeque::new(),
            queue_bytes: 0,
        }
    }

    pub fn enqueue(&mut self, data: Bytes) {
        self.queue_bytes += data.len();
        self.queue.push_back(data);
    }

    pub fn drain(&mut self, loss_estimator: &LossEstimator) -> Vec<Bytes> {
        if self.queue.is_empty() {
            return Vec::new();
        }

        let budget = self.get_budget(loss_estimator);
        let mut sent = 0usize;
        let mut packets = Vec::new();

        while let Some(front) = self.queue.front() {
            if sent + front.len() > budget {
                break;
            }
            let packet = self.queue.pop_front().unwrap();
            sent += packet.len();
            self.queue_bytes -= packet.len();
            packets.push(packet);
        }

        packets
    }

    fn get_budget(&self, loss_estimator: &LossEstimator) -> usize {
        let stats = loss_estimator.get_stats();

        let t = (stats.loss_rate / LOSS_THRESHOLD).min(1.0);
        let target_intervals = MIN_INTERVALS + t * (MAX_INTERVALS - MIN_INTERVALS);
        let adaptive_budget = (self.queue_bytes as f64 / target_intervals) as usize;

        adaptive_budget.max(MIN_BUDGET_BYTES)
    }

    pub fn pending(&self) -> usize {
        self.queue.len()
    }

    pub fn reset(&mut self) {
        self.queue.clear();
        self.queue_bytes = 0;
    }
}

impl Default for PacketPacer {
    fn default() -> Self {
        Self::new()
    }
}
